### PR TITLE
[DS-4082] Replace the term "Entity"

### PR DESCRIPTION
**Pull request by benbosman**
https://jira.duraspace.org/browse/DS-4082

This branch replaces the term "Entity" with a sensible alternative depending on the context.
It's often replaced by "Item" or "ItemRelationship"

### CODE REVIEW

**Comment by tdonohue** [comment]
I'm a bit confused by the usage of both plurals and singulars in this PR, e.g. `ItemRelationshipsType` and `ItemRelationshipsUtil` (both plural), but `ItemRelationshipTypeService` and `ItemRelationshipTypeDAO` (both singular).   We may want to talk more about what to call these.

I also notice that, while variables all are renamed, most of the JavaDocs still uses the term "Entity" throughout.

**Comment by tdonohue** [comment]
This ticket/PR has been delayed until post-7.0 Preview Release. As of today's [Entities WG meeting](https://wiki.duraspace.org/display/DSPACE/2019-01-08+DSpace+7+Entities+WG+Meeting), we've decided we need more feedback from the Marketing Group, etc before moving forward.

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java [hunk 1]
```diff
@@ -23,16 +23,17 @@
...
+public class ItemRelationshipsType implements ReloadableEntity<Integer> {
 
     /**
      * The Integer ID used as a primary key for this database object.
      * This is generated by a sequence
      */
     @Id
-    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "entity_type_id_seq")
-    @SequenceGenerator(name = "entity_type_id_seq", sequenceName = "entity_type_id_seq", allocationSize = 1)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "item_relationships_type_id_seq")
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java]: I would suggest sequence "item_type_id_seq"

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java [hunk 2]
```diff
@@ -23,16 +23,17 @@
...
+    @SequenceGenerator(name = "item_relationships_type_id_seq", sequenceName = "item_relationships_type_id_seq",
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java]: I would suggest sequence "item_type_id_seq"

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java [hunk 3]
```diff
@@ -82,42 +83,46 @@ public EntityType getType(Context context, Entity entity) throws SQLException {
...
-    public List<RelationshipType> getRightRelationshipTypes(Context context, Entity entity) throws SQLException {
-        EntityType entityType = this.getType(context, entity);
+    public List<RelationshipType> getRightRelationshipTypes(Context context,
+                                                            ItemRelationshipsUtil itemRelationshipsUtil)
+        throws SQLException {
+        ItemRelationshipsType itemRelationshipsType = this.getType(context, itemRelationshipsUtil);
         List<RelationshipType> listToReturn = new LinkedList<>();
         for (RelationshipType relationshipType : relationshipTypeService.findAll(context)) {
-            if (relationshipType.getRightType() == entityType) {
+            if (relationshipType.getRightType() == itemRelationshipsType) {
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java]: itemRelationshipsType => itemType

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 4]
```diff
@@ -34,9 +34,9 @@
 import org.dspace.authorize.AuthorizeException;
 import org.dspace.content.Collection;
 import org.dspace.content.DSpaceObject;
-import org.dspace.content.Entity;
-import org.dspace.content.EntityType;
 import org.dspace.content.Item;
+import org.dspace.content.ItemRelationshipsType;
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: ItemRelationshipsType => ItemType

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 5]
```diff
@@ -34,9 +34,9 @@
...
+import org.dspace.content.ItemRelationshipsUtil;
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: This "wrapper class" doesn't have the same behavior as others in this "content" package. Does it make to have it on a different package? 
The class name, I would prefer to call it ItemWrapper or leave as Entity.
As it is, this class enhances an item. If an installer wants, he can choose to keep the default DSpace, upgrade to DS7, leaving the data structure as it is. This makes DSpace more flexible, but perhaps a bit more complex. Why can't we print this behavior directly on the Item class?

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 6]
```diff
@@ -45,9 +45,9 @@
 import org.dspace.content.authority.Choices;
 import org.dspace.content.factory.ContentServiceFactory;
 import org.dspace.content.service.CollectionService;
-import org.dspace.content.service.EntityService;
-import org.dspace.content.service.EntityTypeService;
 import org.dspace.content.service.InstallItemService;
+import org.dspace.content.service.ItemRelationshipTypeService;
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: ItemRelationshipTypeService => ItemTypeService

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipTypeServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+import org.apache.commons.collections.CollectionUtils;
+import org.dspace.authorize.AuthorizeException;
+import org.dspace.authorize.service.AuthorizeService;
+import org.dspace.content.dao.ItemRelationshipTypeDAO;
+import org.dspace.content.service.ItemRelationshipTypeService;
+import org.dspace.core.Context;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class ItemRelationshipTypeServiceImpl implements ItemRelationshipTypeService {
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipTypeServiceImpl.java]: ItemRelationshipTypeServiceImpl  => ItemTypeServiceImpl
ItemRelationshipTypeService => ItemTypeService

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipTypeServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Autowired(required = true)
+    protected ItemRelationshipTypeDAO itemRelationshipTypeDAO;
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipTypeServiceImpl.java]: ItemRelationshipTypeDAO => ItemTypeDAO

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java [hunk 9]
```diff
@@ -13,18 +13,18 @@
...
+import org.dspace.content.service.ItemRelationshipTypeService;
+import org.dspace.content.service.ItemRelationshipsService;
 import org.dspace.content.service.ItemService;
 import org.dspace.content.service.RelationshipService;
 import org.dspace.content.service.RelationshipTypeService;
 import org.dspace.core.Context;
 import org.springframework.beans.factory.annotation.Autowired;
 
-public class EntityServiceImpl implements EntityService {
+public class ItemRelationshipsServiceImpl implements ItemRelationshipsService {
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java]: ItemRelationshipsServiceImpl  => ItemTypeServiceImpl

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java [hunk 10]
```diff
@@ -82,42 +83,46 @@ public EntityType getType(Context context, Entity entity) throws SQLException {
...
-    public List<RelationshipType> getAllRelationshipTypes(Context context, Entity entity) throws SQLException {
-        EntityType entityType = this.getType(context, entity);
+    public List<RelationshipType> getAllRelationshipTypes(Context context, ItemRelationshipsUtil itemRelationshipsUtil)
+        throws SQLException {
+        ItemRelationshipsType itemRelationshipsType = this.getType(context, itemRelationshipsUtil);
         List<RelationshipType> listToReturn = new LinkedList<>();
         for (RelationshipType relationshipType : relationshipTypeService.findAll(context)) {
-            if (relationshipType.getLeftType() == entityType ||
-                relationshipType.getRightType() == entityType) {
+            if (relationshipType.getLeftType() == itemRelationshipsType ||
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/content/ItemRelationshipsServiceImpl.java]: itemRelationshipsType  => itemType

**Review by paulo-graca — COMMENTED** [review]
Summarizing I would prefer simpler names, like:

EntityType = ItemType
ItemRelationshipsType => ItemType
ItemRelationshipTypeService => ItemTypeService
ItemRelationshipTypeServiceImpl => ItemTypeServiceImpl

"Entity": we have to decide if we implement the behavior of this class in the Item class as Tim said in our last Entities meeting, or leave as it is.

File: dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java [hunk 11]
```diff
@@ -23,16 +23,17 @@
  * This also has a label that will be used to identify what kind of EntityType this object is
  */
 @Entity
-@Table(name = "entity_type")
-public class EntityType implements ReloadableEntity<Integer> {
+@Table(name = "item_relationships_type")
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/content/ItemRelationshipsType.java]: I agree with @paulo-graca's suggestion of attempting to use the phrase "Item Type" to replace "Entity Type" (in all of his comments here).

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java (hunk 5 above)
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: I agree with the general comments here that we might want to look into moving this behavior to the Item class itself.  In the current structure, this `ItemRelationshipsUtil` class is an oddity, but if we found a reason to keep it we could call it `ItemRelationshipsWrapper` (as it wraps an Item adding in relationships).

**Comment by tdonohue** [comment]
Closing this PR.  Today the DSpace 7 Marketing Working Group discussed the "Configurable Entities" name/terminology and decided to keep it. See notes at https://wiki.duraspace.org/pages/viewpage.action?pageId=112527588

This means that, at the very least, any additional renaming of classes is no longer a priority.  However, if we feel it adds clarity (for developers, etc), we can still choose to rename classes to use "Item" instead of "Entity".  That would just involve re-opening a new PR (as this one is outdated anyhow).

**Pull request closed without merging**